pyrelational/strategies/generic_al_strategy.py

```diff
@@ -1,293 +1,72 @@
-            else:
-                columns = ["Iteration"]
-            logger.info("COLUMNS: {}".format(columns))
-            logger.info("Full Missing")
-        for k in keys:
-            row = [k]
-            logger.info(self.performances[k])
-            for c in columns[1:]:
-                if isinstance(self.performances[k], list):
-                    row.append(self.performances[k][0][c])
-                else:
-                    row.append(self.performances[k][c])
-            df.append(row)
-
-        pd_df = pd.DataFrame(df, columns=columns)
-        return pd_df
-
-    def log_labelled_by(self, indices: List[int], tag: str = ""):
-        """
-        Update the dictionary that records what the observation
-        was labelled by. Default behaviour is to map observation to
-        iteration at which it was labelled
-
-        :param indices: list of indices selected for labelling
-        :param tag: string which indicates what the observations where labelled by
-        """
-        for indx in indices:
-            self.labelled_by[indx] = tag
-
-    @property
-    def u_indices(self) -> List[int]:
-        return self.data_manager.u_indices
-
-    @property
-    def u_loader(self) -> DataLoader:
-        return self.data_manager.get_unlabelled_loader()
-
-    @property
-    def l_indices(self) -> List[int]:
-        return self.data_manager.l_indices
-
-    @property
-    def l_loader(self) -> DataLoader:
-        return self.data_manager.get_labelled_loader()
-
-    @property
-    def train_loader(self) -> DataLoader:
-        return self.data_manager.get_train_loader(full=True)
-
-    @property
-    def valid_loader(self) -> DataLoader:
-        return self.data_manager.get_validation_loader()
-
-    @property
-    def test_loader(self) -> DataLoader:
-        return self.data_manager.get_test_loader()
-
-    @property
-    def percentage_labelled(self) -> float:
-        return self.data_manager.percentage_labelled()
-
-    @property
-    def dataset_size(self) -> int:
-        return len(self.data_manager)
-
-    def __repr__(self) -> str:
+    def _filter_kwargs(self, **kwargs: Any) -> Dict[str, Any]:
+        """filter kwargs such that they match the active_learning_step signature of the concrete strategy."""
+
+        # filter all parameters based on update signature except those of
+        # type VAR_POSITIONAL (*args) and VAR_KEYWORD (**kwargs)
+        _params = (inspect.Parameter.VAR_POSITIONAL, inspect.Parameter.VAR_KEYWORD)
+        _sign_params = inspect.signature(self.active_learning_step).parameters
+        filtered_kwargs = {
+            k: v for k, v in kwargs.items() if (k in _sign_params.keys() and _sign_params[k].kind not in _params)
+        }
+
+        exists_var_keyword = any(v.kind == inspect.Parameter.VAR_KEYWORD for v in _sign_params.values())
+        # if no kwargs filtered, return all kwargs as default
+        if not filtered_kwargs and not exists_var_keyword:
```

Comment:
If this is true, then filtered_kwargs is already an empty dictionary, so the subsequent line(`filtered_kwargs = {}`) seems redundant to me?

Reply:
Updated in 79dff74